modules/internal/ChapelAutoAggregation.chpl

```diff
@@ -0,0 +1,377 @@
+/*
+ * Copyright 2020-2021 Hewlett Packard Enterprise Development LP
+ * Copyright 2004-2019 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+module ChapelAutoAggregation {
+  private use CommAggregation;
+
+  config param verboseAggregation = false;
+  config param yieldFrequency = 1024;
+  config param dstBuffSize = 4096;
+  config param srcBuffSize = 4096;
+
+  pragma "aggregator generator"
+  proc chpl_srcAggregatorForArr(arr: []) {
+    return new SrcAggregator(arr.eltType);
+  }
+
+  pragma "aggregator generator"
+  proc chpl_srcAggregatorForArr(type t) {
+    return new SrcAggregator(t);
+  }
+
+  pragma "aggregator generator"
+  proc chpl_srcAggregatorForArr(arr) {
+    return false;
+  }
+
+  pragma "aggregator generator"
+  proc chpl_dstAggregatorForArr(arr: []) {
+    return new DstAggregator(arr.eltType);
+  }
+
+  pragma "aggregator generator"
+  proc chpl_dstAggregatorForArr(type t) {
+    return new DstAggregator(t);
+  }
+
+  pragma "aggregator generator"
+  proc chpl_dstAggregatorForArr(arr) {
+    return false;
+  }
+
+  proc chpl__arrayIteratorYieldsLocalElements(a) param {
+    if isArray(a) {
+      if !isClass(a.eltType) { // I have no idea if we can do this for wide pointers
+        return a.iteratorYieldsLocalElements();
+      }
+    }
+    return false;
+  }
+
+  // make sure to resolve this, so that we can give a more meaningful error
+  // message
+  proc chpl__arrayIteratorYieldsLocalElements(type a) param {
+    return false;
+  }
+
+  module CommAggregation {
```

Comment:
Instead of copying the arkouda aggregators, I think  this should copy `test/studies/bale/aggregation/{CopyAggregation.chpl,AggregationPrimitives.chpl}`. I'm find keeping them as submodules in here though. 

AggregationPrimitives.chpl provides building blocks for aggregation and makes it easier to for instance add atomic aggregators. There were probably some cleanups I made to CopyAggregation.chpl as well when copying it from arkouda. 

I would also be in favor of removing those modules from the test directory then, and just having the tests use our internal versions now.

Reply:
> I would also be in favor of removing those modules from the test directory then, and just having the tests use our internal versions now.

I intentionally put them in submodules and `private use`d them, thinking that the user code should never see these modules. Do you want to make that `use` a public one?

I slightly prefer to make that as a separate PR, as it relates more to making aggregators a user-facing feature.

Reply:
I don't think they need to be user-facing just yet, just available to our tests. I just don't want to have 2 different aggregators in our repo and have them get out of sync.

Reply:
> I don't think they need to be user-facing just yet, just available to our tests.

Aren't they the same thing?

We can split this module into two, and put the core of the aggregators in a separate internal module, but not add it in `ChapelStandard`. Have `ChapelAutoAggregation` `private use` it (I think it should be able to because they are in the same path), and symlink to the aggregator module in the test?

Reply:
I'm imaging the tests are just doing something like `use ChapelAutoAggregation.CommAggregation;`

So not really user-facing since it's not documented and it's using internal names, but it's still accessible to our tests.

Reply:
Ah, I see.

I am half-surprised that we could do that. I'll try.